test/k8sT/hubble.go

```diff
@@ -0,0 +1,207 @@
+		node, err := kubectl.GetNodeNameByLabel(label)
+		if err != nil {
+			return "", fmt.Errorf("unable to find node for label %s", label)
+		}
+
+		filter := fmt.Sprintf(
+			"-o jsonpath='{.items[?(@.spec.nodeName == \"%s\")].metadata.name}'", node)
+
+		res := kubectl.ExecShort(fmt.Sprintf(
+			"%s -n %s get pods %s %s", helpers.KubectlCmd, ns, hubbleSelector, filter))
+		if !res.WasSuccessful() {
+			return "", fmt.Errorf("hubble-cli pod not found on node '%s'", node)
+		}
+		return res.Output().String(), nil
+	}
+
+	hubbleExecUntilMatch := func(ns, pod, cmd, substr string) error {
+		By("Executing %q on %s/%s", cmd, ns, pod)
+		body := func() bool {
+			ctx, cancel := context.WithTimeout(context.Background(), helpers.ShortCommandTimeout)
+			defer cancel()
+			res := kubectl.ExecPodCmdContext(ctx, ns, pod, cmd)
+			if !res.WasSuccessful() {
+				return false
+			}
+			return strings.Contains(res.Output().String(), substr)
+		}
+
+		return helpers.WithTimeout(
+			body,
+			fmt.Sprintf("%s is not in the hubble output after timeout", substr),
+			&helpers.TimeoutConfig{Timeout: helpers.MidCommandTimeout})
+	}
+
+	addVisibilityAnnotation := func(ns, podLabels, direction, port, l4proto, l7proto string) {
+		visibilityAnnotation := fmt.Sprintf("<%s/%s/%s/%s>", direction, port, l4proto, l7proto)
+		By("Adding visibility annotation %s on pod with labels %s", visibilityAnnotation, podLabels)
+
+		// Prints <node>=<ns>/<podname> for each pod the annotation was applied to
+		res := kubectl.Exec(fmt.Sprintf("%s annotate pod -n %s -l %s %s=%q"+
+			" -o 'jsonpath={.spec.nodeName}={.metadata.namespace}/{.metadata.name}{\"\\n\"}'",
+			helpers.KubectlCmd,
+			ns, app1Labels,
+			annotation.ProxyVisibility, visibilityAnnotation))
+		res.ExpectSuccess("adding proxy visibility annotation failed")
+
+		// For each pod, check that the Cilium proxy-statistics contain the new annotation
+		expectedProxyState := strings.ToLower(visibilityAnnotation)
+		for node, podName := range res.KVOutput() {
+			ciliumPod, err := kubectl.GetCiliumPodOnNode(helpers.CiliumNamespace, node)
+			Expect(err).To(BeNil())
+
+			// Extract annotation from endpoint model of pod. It does not have the l4proto, so we insert it manually.
+			cmd := fmt.Sprintf("cilium endpoint get pod-name:%s"+
+				" -o jsonpath='{range [*].status.policy.proxy-statistics[*]}<{.location}/{.port}/%s/{.protocol}>{\"\\n\"}{end}'",
+				podName, strings.ToLower(l4proto))
+			err = kubectl.CiliumExecUntilMatch(ciliumPod, cmd, expectedProxyState)
+			Expect(err).To(BeNil(), "timed out waiting for endpoint to regenerate for visibility annotation")
+		}
+	}
+
+	removeVisbilityAnnotation := func(ns, podLabels string) {
+		By("Removing visibility annotation on pod with labels %s", app1Labels)
+		res := kubectl.Exec(fmt.Sprintf("%s annotate pod -n %s -l %s %s-", helpers.KubectlCmd, ns, podLabels, annotation.ProxyVisibility))
+		res.ExpectSuccess("removing proxy visibility annotation failed")
```

Comment:
Do we need to wait here until the annotation is gone?

Reply:
For the current test setup it's not required, since we don't even need to remove the annotation in the first place.

But you bring up a good point. As we will enable Hubble for other e2e tests as well, I plan to move out many functions such as this one into the `helpers` package anyways. I will make sure to add the wait statement there.